HipercowApi/Controllers/ClusterAccessController.cs

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Imperial College London. All rights reserved.
+
+namespace HipercowApi.Controllers
+{
+    using HipercowApi.Tools;
+    using Microsoft.AspNetCore.Authorization;
+    using Microsoft.AspNetCore.Mvc;
+
+    /// <summary>
+    /// The /clusteraccess endpoint requires an authentication
+    /// token, and provides the list of clusters that a certain
+    /// user has access to.
+    /// </summary>
+    [ApiController]
+    [Route("api/v1/[controller]")]
+    public class ClusterAccessController : ControllerBase
+    {
+        private readonly UserSessionManager sessionManager;
+
+        /// <summary>
+        /// Initializes a new instance of the
+        /// <see cref="ClusterAccessController"/> class.
+        /// </summary>
+        /// <param name="sessionManager">The session manager so we can look up the
+        /// login details of previous sessions in the memory cache.
+        /// </param>
+        public ClusterAccessController(
+            UserSessionManager sessionManager)
+        {
+            this.sessionManager = sessionManager;
+        }
+
+        /// <summary>
+        /// Authenticated endpoint to return the list of clusters available to a user.
+        /// </summary>
+        /// <param name="sessionId">Session Id for retrieving login details.</param>
+        /// <returns>
+        /// A list of cluster names.
+        /// </returns>
+        [Authorize]
+        [HttpGet]
+        public IActionResult GetMyClusters([FromHeader(Name = "X-Session-Id")] string sessionId)
+        {
+            string? jwtUsername = this.User.Identity!.Name;
+            UserSession? session = this.sessionManager.RetrieveSession(sessionId);
+            IActionResult? result = Utils.CheckTokenAndSession(this, jwtUsername, session);
+            if (result is not null)
+            {
+                return result;
+            }
+
+            // Use session info:
+            if (session!.Wpia_hn_access)
+            {
+                return this.Ok("wpia-hn");
+            }
+            else
+            {
+                return this.Unauthorized("Domain authentication ok, but no access to any clusters.");
```

Comment:
I don't think this should return a 401 - they've authenticated ok, but they just don't have any cluster access. Should be an empty list?

Reply:
Quick talk with Rich - I think we felt a Forbidden (403) was ok if they authenticate ok, but aren't in any cluster groups, and an Unauthorized (401) if their dide/user was wrong... I think I've got that the right way round now.